### PR TITLE
Identifier Utils Module

### PR DESCRIPTION
**Pull request by ammokhov**
*Issue #, if available:*

*Description of changes:*

added identifier utils module that generates cfn compliant resource name `stackName-logicalId-hash`

Example:
stack-name: `my-custom-stack`
logical id: `myresource`

output: `my-custom-stack-myresource-191b9b1a51f5`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jaymccon** [comment]
would be nice to get a usage example added to `handlers.py` template

File: tests/lib/identifier_utils_test.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_generated_name_with_max_len_shorter_than_preferred():
+    result: str = generate_resource_identifier(
+        stack_id_or_name="abc",
+        logical_resource_id="abc",
+        client_request_token="123456789",
+        max_length=16,
+    )
+    assert len(result) == 16
+    assert result == "aba-f7c3bc1d808e"
```
Comment (jotompki) [inline tests/lib/identifier_utils_test.py:49]: i dont really like asserting on hash values. makes tests very brittle. can we check for a prefix or something instead

File: src/cloudformation_cli_python_lib/identifier_utils.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        raise Exception(
+            f"Cannot generate resource IDs shorter than\
+               {MIN_PHYSICAL_RESOURCE_ID_LENGTH} characters."
+        )
+
+    stack_name: str = stack_id_or_name
+
+    pattern = re.compile(STACK_ARN_PATTERN)
+
+    if pattern.match(stack_id_or_name):
```
Comment (jotompki) [inline src/cloudformation_cli_python_lib/identifier_utils.py:32]: maybe just augment the regex to get a [named group](https://docs.python.org/3/howto/regex.html#non-capturing-and-named-groups). won't have to split the stack arn again to extract the name

File: src/cloudformation_cli_python_lib/identifier_utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import hashlib
+import re
+from typing import List
+
+STACK_ARN_PATTERN = "^[a-z0-9-:]*stack/[-a-z0-9A-Z]*/[-a-z0-9A-Z]*"
+
+MIN_PHYSICAL_RESOURCE_ID_LENGTH = 15
```
Comment (jotompki) [inline src/cloudformation_cli_python_lib/identifier_utils.py:7]: where is this coming from?

File: src/cloudformation_cli_python_lib/identifier_utils.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+MIN_PREFERRED_LENGTH = 17
+HASH_LENGTH = 12
+
+
+def _get_hash(client_request_token: str) -> str:
+    return hashlib.sha1(str.encode(client_request_token)).hexdigest()  # nosec
+
+
+def generate_resource_identifier(
```
Comment (jotompki) [inline src/cloudformation_cli_python_lib/identifier_utils.py:16]: is this trying to achieve parity with some existing naming scheme?

File: src/cloudformation_cli_python_lib/identifier_utils.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    pattern = re.compile(STACK_ARN_PATTERN)
+
+    if pattern.match(stack_id_or_name):
+        stack_name = stack_name.split("/")[1]
+
+    separate: bool = max_length > MIN_PREFERRED_LENGTH
+
+    clean_stack_name: str = stack_name.replace("^-+", "", 1).replace("--", "-")
+    free_chars: int = max_length - 13 - (1 if separate else 0)
```
Comment (jotompki) [inline src/cloudformation_cli_python_lib/identifier_utils.py:38]: lets make 13 a constant variable

File: src/cloudformation_cli_python_lib/identifier_utils.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    requested_lengths: List[int] = [len(clean_stack_name), len(logical_resource_id)]
+
+    available_lengths: List[int] = fair_split(free_chars, requested_lengths)
```
Comment (jotompki) [inline src/cloudformation_cli_python_lib/identifier_utils.py:41]: can't we just do something like:
```
chars_for_stack_name = min(free_chars / 2, len(stack_name))
chars_for_resource_name = min(free_chars - chars_for_stack_name, len(resource_name)
```

and ditch the fair split method?

**Review by jotompki — CHANGES_REQUESTED** [review]

File: src/cloudformation_cli_python_lib/identifier_utils.py (hunk 5 above)
Comment (ammokhov) [reply src/cloudformation_cli_python_lib/identifier_utils.py:38]: will do

File: tests/lib/identifier_utils_test.py (hunk 1 above)
Comment (ammokhov) [reply tests/lib/identifier_utils_test.py:49]: will do

File: src/cloudformation_cli_python_lib/identifier_utils.py (hunk 5 above)
Comment (jotompki) [reply src/cloudformation_cli_python_lib/identifier_utils.py:39]: still not done

**Review by jotompki — APPROVED** [review]
Everything else looks good other than the 13 issue

**Review by omkhegde — APPROVED** [review]

**Pull request merged**